##########################################################
# Parameter Configurations for TMCM-1617
##########################################################

# Note: Trinamic Motor Controllers has a maximum support of 3 axes/motors, thus the vector for each setting consists of 3 elements.

########## Interface Settings

# Communication interface
# Description: Interface used between the PC and the TMC
# Possible values:
# - 0: CAN
comm_interface: 0

# Communication interface name
# Description: Name of the interface or device as detected by the PC
# Possible values:
# - can0, can1, can2, ..., etc
comm_interface_name: can0

# Communication interface bit rate
# Description: Rate of communication interface (bits per sec)
# Possible values:
# - 20000
# - 50000
# - 100000
# - 125000
# - 250000
# - 500000
# - 1000000
# Note: The bit rate that would work will be the one set at the TMC's GP: 69
comm_bit_rate: 1000000

# Communication interface Tx ID
# Description: Currently applicable only for CAN interface (Tx ID)
# Possible values:
# - 0 to 255
# Note: The ID that would work will be the one set at the TMC's GP: 71
comm_tx_id: 1

# Communication interface Rx ID
# Description: Currently applicable only for CAN interface (Rx ID)
# Possible values:
# - 0 to 255
# Note: The ID that would work will be the one set at the TMC's GP: 70
comm_rx_id: 2

# Communication timeout rate in seconds
# Description: Indicates how long should the node will wait for the rx data
# Possible values:
# - value in millisecs 0 - 5000
comm_timeout_ms: 100

# Communication timeout tries
# Description: Indicates how many the node will retry sending data before shutting down if no data received
# Possible values:
# - 1 - 3
comm_exec_cmd_retries: 1

########## Additional Settings

# tmc_info topics to publish into
# Description: tmc_info topics that will contain chosen TMC info that will be published
# Possible values:
# - Any tmc_info topic name
tmc_info_topic: [/tmc_info_0, /tmc_info, /tmc_info]

# cmd_vel topics to subscribe from
# Description: Twist topics that will be the source of target velocity to be set on the TMC
# Possible values: 
# - Any Twist topic name
# Note: Only its linear.x will be used (range: -200000 to 200000)
tmc_cmd_vel_topic: [/cmd_vel_0, /cmd_vel, /cmd_vel]

# cmd_pos topics to subscribe from
# Description: Int32 topics that will be the source of target position to be set on the TMC
# Possible values: 
# - Any Int32 topic name (range: -2147483648 to -2147483647)
tmc_cmd_abspos_topic: [/cmd_abspos_0, /cmd_abspos, /cmd_abspos]

# cmd_pos topics to subscribe from
# Description: Int32 topics that will be the source of target position to be set on the TMC
# Possible values: 
# - Any Int32 topic name (range: -2147483648 to -2147483647)
tmc_cmd_relpos_topic: [/cmd_relpos_0, /cmd_relpos, /cmd_relpos]

# cmd_trq topics to subscribe from
# Description: Int32 topics that will be the source of target torque to be set on the TMC
# Possible values: 
# - Any Int32 topic name (range: -60000 to 60000 mA)
tmc_cmd_trq_topic: [/cmd_trq_0, /cmd_trq, /cmd_trq]

# Wheel diameter attached to the motor (in  meters)
# Description: Wheel diameter that is attached on the motor. This is to convert linear values to rpm
# Note: If wheel diameter is 0, cmd_vel is equal to rpm
wheel_diameter: 0.25

# Active motors or axes
# Description: Enables/disables active motors or axes. If disabled, settings for those motors will be ignored or not set.
# Possible values:
# - 0: Not used
# - 1: Active
en_motors: [1, 0, 0]

# Publish TMC information
# Decription: Enables/disables publishing of TMC information
# Possible values:
# - true: Enabled
# - false: Disabled
en_pub_tmc_info: [true, false, false] 

# Publish rate of TMC information
# Decription: Publish rate (hertz) of TMC information
# Possible values: 10 - 100
pub_rate_tmc_info: 10

########## TMC Information

# Enable/Disable axis parameters that the user can optionally publish every publish rate as long as en_pub_tmc_info is true

pub_actual_vel: [true, false, false] 
pub_actual_trq: [true, false, false]
pub_actual_pos: [true, false, false]

##########################################################
##########################################################

# The configurations below are the values set for motors 0, 1 and 2 as [<value of motor 0>, <value of motor 1>, <value of motor 2>].
# If some of the motors or configurations are not used, just indicate 0 as its value.
# To know the possible values for all configurations below (eg, "CommutationMode_Ext"),
# check the corresponding configuration in the autogenerated TMCM-XXXX.yaml inside config/autogenerated folder (eg, "CommutationMode").
# All configurations must have a corresponding <Configuration>_Ext values below EXCEPT:
# - SupplyVoltage, TargetVelocity, ActualVelocity, TargetPosition, ActualPosition, TargetTorque and ActualTorque

# Follow EEPROM configurations
# Description: Flag for the ROS Driver to identify where to get motor configurations
# Possible values:
# - true:  Follow all the motor configuration as-is (EEPROM values set as part of TMC firmware or as set in TMCL-IDE)
# - false:  Have user set all motor configurations in this YAML (below)
FollowEepromConfig: true

########## Sensor Selection Settings

# Commutation sensor selection
# Description: Commutation sensor selection
CommutationMode_Ext: [3, 0, 0]

# Velocity sensor selection
# Description: Velocity sensor selection
CommutationModeVelocity_Ext: [0, 0, 0]

# Position sensor selection
# Description: Position sensor selection
CommutationModePosition_Ext: [0, 0, 0]

########## Motor Settings
# Max current (RMS)
# Description: Maximum allowed absolute motor current
MaxTorque_Ext: [4000, 0, 0]

# Open loop current (RMS)
# Description: Motor current for controlled commutation (used if CommutationMode_Ext = 1)
OpenLoopCurrent_Ext: [2000, 0, 0]

# Acceleration(rpm)
# Description: Motor acceleration value
Acceleration_Ext: [50, 0, 0]

# Motor pole pairs
# Description: Number of motor poles
MotorPolePairs_Ext: [4, 0, 0]

# Motor PWM frequency (Hertz)
# Description: Frequency of the motor PWM
PWMFrequency_Ext: [25000, 0, 0]


########## Hall Settings

# Hall polarity
# Description: Hall sensor polarity
HallSensorPolarity_Ext: [1, 0, 0]

# Hall direction
# Description: Hall sensor direction 
HallSensorDirection_Ext: [0, 0, 0]

# Hall interpolation 
# Description: Hall sensor interpolation 
HallInterpolation_Ext: [0, 0, 0]

# Hall PHI_E offset
# Description: Offset for electrical angle hall_phi_e of hall sensor
HallSensorOffset_Ext: [0, 0, 0]

########## Encoder Settings

# Encoder direction
# Description: Encoder direction in a way that ROR increases position counter
EncoderDirection_Ext: [1, 0, 0]

# Encoder steps
# Description: Encoder steps per full motor rotation
EncoderSteps_Ext: [4096, 0, 0]

# Encoder init mode
# Description: Encoder init mode
EncoderInitMode_Ext: [0, 0, 0]

########## PID Settings

# Torque P
# Description: P parameter for current PID regulator
TorqueP_Ext: [300, 0, 0]

# Torque I
# Description: I parameter for current PID regulator
TorqueI_Ext: [500, 0, 0]

# Velocity P
# Description: P parameter for velocity PID regulator
VelocityP_Ext: [400, 0, 0]

# Velocity I
# Description: I parameter for velocity PID regulator
VelocityI_Ext: [100, 0, 0]

# Position P
# Description: P parameter for position PID regulator
PositionP_Ext: [10, 0, 0]

########## Brake Chopper Settings

# Enable brake chopper
# Description: Enable brake chopper fuctionality
BrakeChopperEnabled_Ext: [0, 0, 0]

# Brake chopper voltage limit
# Description: If the brake chopper is enabled and suppyly voltage exceeds this value, the brake chopper output will be activated
BrakeChopperVoltage_Ext: [0, 0, 0]

# Brake chopper hysteresis
# Description: An activated brake chopper will be disabled if the actual supply voltage is lower than limit voltage hysteresis
BrakeChopperHysteresis_Ext: [0, 0, 0]

########## Position Settings

# Position scaler
# Description: Scale the external position (external position = (internal_position * scaler) / 65536)
PositionScalerM_Ext: [65536, 0, 0]

